File: Sources/Speak/Whisper/WhisperContext.swift
import Foundation
import CWhisper

enum WhisperError: Error, LocalizedError {
    case modelLoadFailed(String)
    case transcriptionFailed(Int32)

    var errorDescription: String? {
        switch self {
        case .modelLoadFailed(let path):
            return "Failed to load whisper model at: \(path)"
        case .transcriptionFailed(let code):
            return "Whisper transcription failed with code: \(code)"
        }
    }
}

actor WhisperContext {
    private let ctx: OpaquePointer
    private let settings: WhisperSettings
    private let modelName: String

    init(modelPath: String, settings: WhisperSettings = WhisperSettings()) throws {
        var cparams = whisper_context_default_params()
        cparams.use_gpu = settings.useGPU
        cparams.flash_attn = settings.flashAttention

        guard let context = whisper_init_from_file_with_params(modelPath, cparams) else {
            throw WhisperError.modelLoadFailed(modelPath)
        }
        self.ctx = context
        self.settings = settings
        self.modelName = URL(fileURLWithPath: modelPath).deletingPathExtension().lastPathComponent
    }

    deinit {
        whisper_free(ctx)
    }

    func warmup() {
        NSLog("[WhisperContext] Warming up model...")
        let start = CFAbsoluteTimeGetCurrent()
        let silence = [Float](repeating: 0, count: 16000)
        _ = transcribe(samples: silence)
        let elapsed = (CFAbsoluteTimeGetCurrent() - start) * 1000
        NSLog("[WhisperContext] Warmup complete (%.0fms)", elapsed)
    }

    func transcribe(samples: [Float], contextPrompt: String? = nil) -> TranscriptionResult {
        let start = CFAbsoluteTimeGetCurrent()

        var params = whisper_full_default_params(settings.strategy.whisperStrategy)

        params.n_threads = Int32(settings.resolvedThreadCount)
        params.translate = settings.translate
        params.no_context = contextPrompt == nil ? settings.noContext : false
        params.no_timestamps = settings.noTimestamps
        params.single_segment = settings.singleSegment
        params.token_timestamps = settings.tokenTimestamps
        params.suppress_blank = settings.suppressBlank
        params.suppress_nst = settings.suppressNonSpeechTokens
        params.temperature = settings.temperature
        params.temperature_inc = settings.temperatureInc
        params.carry_initial_prompt = settings.carryInitialPrompt
        params.entropy_thold = settings.entropyThreshold
        params.logprob_thold = settings.logprobThreshold
        params.no_speech_thold = settings.noSpeechThreshold
        params.greedy.best_of = Int32(settings.bestOf)
        params.beam_search.beam_size = Int32(settings.beamSize)
        params.print_special = false
        params.print_progress = false
        params.print_realtime = false
        params.print_timestamps = false

        let languageCStr = settings.language.withCString { strdup($0) }
        params.language = UnsafePointer(languageCStr)
        defer { free(languageCStr) }

        let prompt = contextPrompt ?? (settings.initialPrompt.isEmpty ? nil : settings.initialPrompt)
        let promptCStr: UnsafeMutablePointer<CChar>?
        if let prompt = prompt {
            promptCStr = prompt.withCString { strdup($0) }
            params.initial_prompt = UnsafePointer(promptCStr)
        } else {
            promptCStr = nil
            params.initial_prompt = nil
        }
        defer { free(promptCStr) }

        let vadModelPath = ModelManager.vadModelPath
        let vadCStr: UnsafeMutablePointer<CChar>?
        if settings.sileroVADEnabled && ModelManager.vadModelExists {
            params.vad = true
            vadCStr = vadModelPath.withCString { strdup($0) }
            params.vad_model_path = UnsafePointer(vadCStr)
        } else {
            vadCStr = nil
        }
        defer { free(vadCStr) }

        let result = samples.withUnsafeBufferPointer { buffer in
            whisper_full(ctx, params, buffer.baseAddress, Int32(buffer.count))
        }

        let elapsed = (CFAbsoluteTimeGetCurrent() - start) * 1000.0
        let audioDurationMs = Double(samples.count) / 16.0

        guard result == 0 else {
            return TranscriptionResult(
                segments: [],
                audioDurationMs: audioDurationMs,
                transcriptionTimeMs: elapsed,
                modelName: modelName
            )
        }

        let nSegments = whisper_full_n_segments(ctx)
        var segments: [TranscriptionSegment] = []
        segments.reserveCapacity(Int(nSegments))

        for i in 0..<nSegments {
            let text: String
            if let cStr = whisper_full_get_segment_text(ctx, i) {
                text = String(cString: cStr)
            } else {
                text = ""
            }
            let t0 = whisper_full_get_segment_t0(ctx, i) * 10
            let t1 = whisper_full_get_segment_t1(ctx, i) * 10

            let noSpeechProb = whisper_full_get_segment_no_speech_prob(ctx, i)

            let nTokens = whisper_full_n_tokens(ctx, i)
            var tokenProbSum: Float = 0
            for j in 0..<nTokens {
                tokenProbSum += whisper_full_get_token_p(ctx, i, j)
            }
            let avgTokenProb = nTokens > 0 ? tokenProbSum / Float(nTokens) : 0

            segments.append(TranscriptionSegment(
                text: text,
                startTime: t0,
                endTime: t1,
                noSpeechProb: noSpeechProb,
                avgTokenProb: avgTokenProb
            ))
        }

        return TranscriptionResult(
            segments: segments,
            audioDurationMs: audioDurationMs,
            transcriptionTimeMs: elapsed,
            modelName: modelName
        )
    }
}

File: linux/src/whisper_context.cpp
#include "whisper_context.h"
#include "whisper.h"
#include <chrono>
#include <stdexcept>
#include <filesystem>
#include <cstring>

WhisperContext::WhisperContext(const std::string& model_path, const Settings& settings)
    : settings_(settings) {

    auto cparams = whisper_context_default_params();
    cparams.use_gpu = settings.use_gpu;
    cparams.flash_attn = settings.flash_attention;

    ctx_ = whisper_init_from_file_with_params(model_path.c_str(), cparams);
    if (!ctx_) throw std::runtime_error("Failed to load whisper model: " + model_path);

    model_name_ = std::filesystem::path(model_path).stem().string();
}

WhisperContext::~WhisperContext() {
    if (ctx_) whisper_free(ctx_);
}

void WhisperContext::warmup() {
    fprintf(stderr, "[WhisperContext] Warming up model...\n");
    auto start = std::chrono::steady_clock::now();
    std::vector<float> silence(16000, 0.0f);
    transcribe(silence);
    auto elapsed = std::chrono::duration<double, std::milli>(
        std::chrono::steady_clock::now() - start).count();
    fprintf(stderr, "[WhisperContext] Warmup complete (%.0fms)\n", elapsed);
}

TranscriptionResult WhisperContext::transcribe(const std::vector<float>& samples, const std::string* context_prompt) {
    std::lock_guard<std::mutex> lk(mu_);

    auto start = std::chrono::steady_clock::now();

    auto params = whisper_full_default_params(
        settings_.strategy == SamplingStrategy::beam_search
            ? WHISPER_SAMPLING_BEAM_SEARCH
            : WHISPER_SAMPLING_GREEDY);

    params.n_threads = settings_.resolved_thread_count();
    params.translate = settings_.translate;
    params.no_context = (context_prompt == nullptr) ? settings_.no_context : false;
    params.no_timestamps = settings_.no_timestamps;
    params.single_segment = settings_.single_segment;
    params.token_timestamps = settings_.token_timestamps;
    params.suppress_blank = settings_.suppress_blank;
    params.suppress_nst = settings_.suppress_non_speech_tokens;
    params.temperature = settings_.temperature;
    params.entropy_thold = settings_.entropy_threshold;
    params.logprob_thold = settings_.logprob_threshold;
    params.no_speech_thold = settings_.no_speech_threshold;
    params.greedy.best_of = settings_.best_of;
    params.beam_search.beam_size = settings_.beam_size;
    params.print_special = false;
    params.print_progress = false;
    params.print_realtime = false;
    params.print_timestamps = false;

    params.language = settings_.language.c_str();

    const std::string* prompt = context_prompt;
    std::string fallback;
    if (!prompt && !settings_.initial_prompt.empty()) {
        fallback = settings_.initial_prompt;
        prompt = &fallback;
    }
    params.initial_prompt = prompt ? prompt->c_str() : nullptr;

    int result = whisper_full(ctx_, params, samples.data(), static_cast<int>(samples.size()));

    auto elapsed = std::chrono::duration<double, std::milli>(
        std::chrono::steady_clock::now() - start).count();
    double audio_ms = static_cast<double>(samples.size()) / 16.0;

    TranscriptionResult tr;
    tr.audio_duration_ms = audio_ms;
    tr.transcription_time_ms = elapsed;
    tr.model_name = model_name_;

    if (result != 0) return tr;

    int n_segments = whisper_full_n_segments(ctx_);
    tr.segments.reserve(n_segments);

    for (int i = 0; i < n_segments; ++i) {
        const char* text = whisper_full_get_segment_text(ctx_, i);
        int64_t t0 = whisper_full_get_segment_t0(ctx_, i) * 10;
        int64_t t1 = whisper_full_get_segment_t1(ctx_, i) * 10;
        tr.segments.push_back({text ? text : "", t0, t1});
    }

    return tr;
}
